### PR TITLE
#4338 pg schema upgrade

### PR DESCRIPTION
**Pull request by paulovitin**
Fix this problem #4338 

When the Postgres Storage Adapter perform the initialization, and the "Volatile Classes Schemas" perform the table creation, the schemaUpgrade method execute a diff between database columns and the Schemas in code. If has new columns, create then through method addFieldIfNotExists.

### CODE REVIEW

**Comment by montymxb** [comment]
I just retriggered the build, should be ok. I still need to compare this against the actual upgrade issue and verify that it does in fact rectify that issue.

File: spec/PostgresStorageAdapter.spec.js [hunk 1]
```diff
@@ -19,4 +30,29 @@ describe_only_db('postgres')('PostgresStorageAdapter', () => {
...
+      fields: {
+        "pushTime": { type: 'String' },
+        "source": { type: 'String' }, // rest or webui
+        "query": { type: 'String' }, // the stringified JSON query
+      },
+    };
+
+    adapter.createTable(className, schema)
+      .then(() => {
+        schema.fields.expiration_interval = { type:'Number' };
```
Comment (montymxb) [inline spec/PostgresStorageAdapter.spec.js]: Just to be pedantic can we** getColumns** and check before the schema is upgraded to ensure that it's not present before hand?

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 2]
```diff
@@ -780,7 +806,12 @@ export class PostgresStorageAdapter {
     return Promise.resolve()
       .then(() => {
         fieldNames = fieldNames.reduce((list, fieldName) => {
-          const field = schema.fields[fieldName]
+          if (!{}.hasOwnProperty.call(schema.fields, fieldName)) {
+            list.push(fieldName);
+            return list;
+          }
+
+          const field = schema.fields[fieldName];
```
Comment (montymxb) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Can you elaborate on this change here?

Comment (paulovitin) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: I dont understand what you said with "elaborate on this change here".. :)

Comment (montymxb) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Hmm, I'll reword it. Could you explain the reason you made the changes above here? I'm not entirely sure why you added this at first glance.

Comment (montymxb) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Specifically what about the prior functionality without `...!{}.hasOwnProperty.call(schema.fields, fieldName)..` wasn't working and needed to be changed?

**Comment by paulovitin** [comment]
@montymxb I need to do something here?

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 3]
```diff
@@ -1577,12 +1603,14 @@ export class PostgresStorageAdapter {
   performInitialization({ VolatileClassesSchemas }) {
     debug('performInitialization');
     const promises = VolatileClassesSchemas.map((schema) => {
-      return this.createTable(schema.className, schema).catch((err) => {
-        if (err.code === PostgresDuplicateRelationError || err.code === Parse.Error.INVALID_CLASS_NAME) {
-          return Promise.resolve();
-        }
-        throw err;
-      });
+      return this.createTable(schema.className, schema)
```
Comment (dplewis) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: You could use transactions here. There are a few examples in the adapter. 
```
return this._client.tx(t => {
  const q1 = this.createTable(schema.className, schema, t).catch(...);
  const q2 = this.schemaUpgrade(schema.className, schema, t);

  return t.batch([q1, q2]);
})
```

**Comment by paulovitin** [comment]
@montymxb How are we staying here?

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 4]
```diff
@@ -745,6 +745,32 @@ export class PostgresStorageAdapter {
       });
   }
 
+  schemaUpgrade(className, schema) {
+    debug('schemaUpgrade', { className, schema });
+    return this._client.tx('schemaUpgrade', t =>
+      t.any(`SELECT column_name FROM information_schema.columns WHERE table_name = '${className}'`)
```
Comment (vitaly-t) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: do not format the values manually, use the query-formatting engine for it

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 5]
```diff
@@ -745,6 +745,32 @@ export class PostgresStorageAdapter {
...
+    )
+      .then(columns => {
+        if (!columns) {
+          return Promise.resolve();
```
Comment (vitaly-t) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: this condition will never happen, see the API for method `any`.

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 6]
```diff
@@ -745,6 +745,32 @@ export class PostgresStorageAdapter {
...
+        }
+
+        const currentColumns = columns.map(item => item.column_name);
```
Comment (vitaly-t) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: perhaps should just use method `map` of the database instead.

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 7]
```diff
@@ -745,6 +745,32 @@ export class PostgresStorageAdapter {
...
+        const newColumns = Object.keys(schema.fields);
+
+        const columnsToCreate = newColumns.filter(item => currentColumns.indexOf(item) === -1);
+
+        const promise = [];
+
+        columnsToCreate.forEach(fieldName => {
+          const type = schema.fields[fieldName];
+          promise.push(this.addFieldIfNotExists(className, fieldName, type));
```
Comment (vitaly-t) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: this is bad connection usage, must use a transaction here.

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 8]
```diff
@@ -745,6 +745,32 @@ export class PostgresStorageAdapter {
...
+        });
+
+        return Promise.all(promise);
```
Comment (vitaly-t) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Never use `Promise.all` on database operations, see: https://github.com/vitaly-t/pg-promise/wiki/Common-Mistakes#tasks-versus-rootdirect-queries

**Comment by paulovitin** [comment]
@vitaly-t I updated the code. Can you take a look?

**Comment by montymxb** [comment]
@paulovitin thanks for getting the coverage up! I'll take a look at these changes now...

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 9]
```diff
@@ -747,6 +747,33 @@ export class PostgresStorageAdapter {
       });
   }
 
+  schemaUpgrade(className, schema) {
+    debug('schemaUpgrade', { className, schema });
+
+    return this._client.tx('schemaUpgrade', t => {
+      return t.any('SELECT column_name FROM information_schema.columns WHERE table_name = $<className>', { className })
+        .then(columns => {
+          const currentColumns = columns.map(item => item.column_name);
```
Comment (montymxb) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: So I see this is the only bit you didn't change here. As is this is passing our tests and looks good. Perhaps it's fine as is. @vitaly-t do you think this is important enough to change or are we good as is?

**Comment by paulovitin** [comment]
@montymxb ok. If needs more changes, let me know

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js (hunk 9 above)
Comment (dplewis) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: @vitaly-t Ding? Zing? Ping?

File: spec/PostgresStorageAdapter.spec.js [hunk 10]
```diff
@@ -1,6 +1,17 @@
 const PostgresStorageAdapter = require('../src/Adapters/Storage/Postgres/PostgresStorageAdapter');
 const databaseURI = 'postgres://localhost:5432/parse_server_postgres_adapter_test_database';
 
+const getColumns = (client, className) => {
+  return client.any('SELECT column_name FROM information_schema.columns WHERE table_name = $<className>', { className })
+    .then(columns => {
+      if (!columns.length) {
```
Comment (vitaly-t) [inline spec/PostgresStorageAdapter.spec.js]: What you do in this `.then` block is pointless, i.e. if the condition matches, the you already end up with an empty array.

File: spec/PostgresStorageAdapter.spec.js [hunk 11]
```diff
@@ -1,6 +1,17 @@
 const PostgresStorageAdapter = require('../src/Adapters/Storage/Postgres/PostgresStorageAdapter');
 const databaseURI = 'postgres://localhost:5432/parse_server_postgres_adapter_test_database';
 
+const getColumns = (client, className) => {
```
Comment (vitaly-t) [inline spec/PostgresStorageAdapter.spec.js]: This whole method can be implemented with a single line, using method [map](http://vitaly-t.github.io/pg-promise/Database.html#map).

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js (hunk 9 above)
Comment (vitaly-t) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: This code should be replaced with ES6 generators, the way it was done for all tasks and transaction in the code.

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 12]
```diff
@@ -747,6 +747,33 @@ export class PostgresStorageAdapter {
       });
   }
 
+  schemaUpgrade(className, schema) {
+    debug('schemaUpgrade', { className, schema });
+
+    return this._client.tx('schemaUpgrade', t => {
+      return t.any('SELECT column_name FROM information_schema.columns WHERE table_name = $<className>', { className })
```
Comment (vitaly-t) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Method [map](http://vitaly-t.github.io/pg-promise/Database.html#map) should be used here.

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 13]
```diff
@@ -747,6 +747,33 @@ export class PostgresStorageAdapter {
...
+          const columnsToCreate = newColumns.filter(item => currentColumns.indexOf(item) === -1);
+
+          const addFields = [];
+
+          columnsToCreate.forEach(fieldName => {
+            const type = schema.fields[fieldName];
+            addFields.push(this.addFieldIfNotExists(className, fieldName, type));
+          });
+
+          if (!addFields.length) {
```
Comment (vitaly-t) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Inefficient use of promises here. Can simply skip returning anything for the same effect.

**Comment by vitaly-t** [comment]
In all, this needs a few updates + conflicts resolved.

**Comment by vitaly-t** [comment]
Correction. No, it needs a lot of rewriting, especially in the light of the conflicts.

I have tried to update it, but the conflicts with the later changes stopped me from being able to finish.

**Comment by paulovitin** [comment]
@vitaly-t conflicts resolved

**Comment by vitaly-t** [comment]
Cool. To avoid further confusion, I will refactor the code now. Please refrain from making changes in the nest 15 mins ;)

**Comment by vitaly-t** [comment]
Ok, I'm done here, for now :)

**Comment by vitaly-t** [comment]
@dplewis I've done my part here, do you want to review it? :smile:

**Comment by dplewis** [comment]
LGTM!

**Comment by montymxb** [comment]
@vitaly-t absolutely 👍 . Thanks for taking a look, postgres via node isn't exactly one of my strong suites 😆 .

File: spec/PostgresStorageAdapter.spec.js (hunk 11 above)
Comment (montymxb) [reply spec/PostgresStorageAdapter.spec.js]: This is a pretty simple one. We should setup map and save some lines :wink:. You've already done quite a lot so let me know if you can get around to this @paulovitin . If not I can put in an additional commit myself.

File: spec/PostgresStorageAdapter.spec.js [hunk 14]
```diff
@@ -19,4 +30,100 @@ describe_only_db('postgres')('PostgresStorageAdapter', () => {
     expect(adapter._client.$pool.ending).toEqual(true);
     done();
   });
+
+  it('schemaUpgrade, upgrade the database schema when schema change', done => {
```
Comment (montymxb) [inline spec/PostgresStorageAdapter.spec.js]: Just a typo
> upgrade the database schema when **the** schema **changes**

File: spec/PostgresStorageAdapter.spec.js [hunk 15]
```diff
@@ -19,4 +30,100 @@ describe_only_db('postgres')('PostgresStorageAdapter', () => {
...
+        expect(columns).toContain('pushTime');
+        expect(columns).toContain('source');
+        expect(columns).toContain('query');
+        expect(columns).toContain('expiration_interval');
+        done();
+      })
+      .catch(error => done.fail(error));
+  });
+
+  it('schemaUpgrade, matain correct schema', done => {
```
Comment (montymxb) [inline spec/PostgresStorageAdapter.spec.js]: typo here too
> **maintains** correct schema

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 16]
```diff
@@ -1579,14 +1595,17 @@ export class PostgresStorageAdapter {
   }
 
   performInitialization({ VolatileClassesSchemas }) {
+    // TODO: This method needs to be rewritten to make proper use of connections (@vitaly-t)
```
Comment (montymxb) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Regarding this TODO we'll want to at least rework this in with transactions as mentioned below.

Comment (vitaly-t) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: I will do it :wink:

Comment (vitaly-t) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: DONE!

Comment (vitaly-t) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: P.S. As for the last item, the `TODO` thing, it definitely should done as a separate PR.

**Comment by vitaly-t** [comment]
A-a-and we're back into conflicts :cry:

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js (hunk 16 above)
Comment (montymxb) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Noted! @paulovitin you're good for that one then, we'll address that down the road.

**Comment by montymxb** [comment]
Hmm, probably related to #4349 being recently merged?

**Comment by flovilmart** [comment]
@vitaly-t yup, intoduced types for the DB adapters and I had many conflicts to survive as well :)

**Comment by vitaly-t** [comment]
Well, I volunteer @paulovitin to resolve the conflicts yet again 😂 I'm not doing it 😂

**Comment by dplewis** [comment]
@vitaly-t  we have a very diverse team. New Years Goal: let’s get the team together sometime. Florent set it up. Let’s go to a conference or something

**Comment by flovilmart** [comment]
@vitaly-t no need to do it! Just fixed them from the phone :)

**Comment by flovilmart** [comment]
We have plenty of contrived on open collective, we could probab’y do something to help everyone get there :)

**Comment by dplewis** [comment]
@flovilmart Make it happen captain

**Comment by vitaly-t** [comment]
I did re-run of those flimsy tests a number of times till they finally all came through :joy:

File: spec/PostgresStorageAdapter.spec.js [hunk 17]
```diff
@@ -1,6 +1,10 @@
 import PostgresStorageAdapter from '../src/Adapters/Storage/Postgres/PostgresStorageAdapter';
 const databaseURI = 'postgres://localhost:5432/parse_server_postgres_adapter_test_database';
 
+const getColumns = (client, className) => {
+  return client.map('SELECT column_name FROM information_schema.columns WHERE table_name = $<className>', { className }, a => a.column_name);
+};
+
```
Comment (montymxb) [inline spec/PostgresStorageAdapter.spec.js]: Didn't realize the above was finally changed, alright!

File: spec/PostgresStorageAdapter.spec.js [hunk 18]
```diff
@@ -19,4 +23,100 @@ describe_only_db('postgres')('PostgresStorageAdapter', () => {
     expect(adapter._client.$pool.ending).toEqual(true);
     done();
   });
+
+  it('schemaUpgrade, upgrade the database schema when schema changes', done => {
```
Comment (montymxb) [inline spec/PostgresStorageAdapter.spec.js]: typo fixed...

File: spec/PostgresStorageAdapter.spec.js [hunk 19]
```diff
@@ -19,4 +23,100 @@ describe_only_db('postgres')('PostgresStorageAdapter', () => {
...
+        expect(columns).toContain('pushTime');
+        expect(columns).toContain('source');
+        expect(columns).toContain('query');
+        expect(columns).toContain('expiration_interval');
+        done();
+      })
+      .catch(error => done.fail(error));
+  });
+
+  it('schemaUpgrade, maintain correct schema', done => {
```
Comment (montymxb) [inline spec/PostgresStorageAdapter.spec.js]: typo fixed...

**Comment by montymxb** [comment]
Thanks again @paulovitin for putting up the PR!

**Pull request merged**